File: server/models/Property.js
const mongoose = require('mongoose');
const slugify = require('slugify');
const geocoder = require('../utils/hybridGeocoder');

console.log('🔧 Property model loaded');
console.log('🔧 Mongoose connection state:', mongoose.connection.readyState);
console.log('🔧 Mongoose version:', mongoose.version);

const PropertySchema = new mongoose.Schema({
  // Basic Information
  title: {
    type: String,
    required: [true, 'Please add a title'],
    trim: true,
    maxlength: [100, 'Title cannot be more than 100 characters']
  },
  slug: String,
  description: {
    type: String,
    required: [true, 'Please add a description'],
    maxlength: [2000, 'Description cannot be more than 2000 characters']
  },
  type: {
    type: String,
    required: [true, 'Please select a property type'],
    enum: [
      'House',
      'Apartment',
      'Villa',
      'Condo',
      'Townhouse',
      'Land',
      'Commercial',
      'PG',
      'Studio',
      'Penthouse',
      'Builder Floor',
      'Farm House',
      'Service Apartment'
    ],
    default: 'House'
  },
  status: {
    type: String,
    required: [true, 'Please select a status'],
    enum: [
      'For Sale',
      'For Rent',
      'Sold',
      'Rented'
    ],
    default: 'For Sale'
  },

  // Pricing and Details
  price: {
    type: Number,
    required: [true, 'Please add a price'],
    min: [0, 'Price must be at least 0']
  },
  bedrooms: {
    type: Number,
    required: [true, 'Please add number of bedrooms'],
    min: [0, 'Bedrooms must be at least 0']
  },
  bathrooms: {
    type: Number,
    required: [true, 'Please add number of bathrooms'],
    min: [0, 'Bathrooms must be at least 0']
  },
  area: {
    type: Number,
    required: [true, 'Please add area in sqft'],
    min: [0, 'Area must be at least 0']
  },
  buildingName: {
    type: String,
    default: '',
    maxlength: [100, 'Building name cannot be more than 100 characters']
  },
  floorNumber: {
    type: String,
    default: '',
    maxlength: [20, 'Floor number cannot be more than 20 characters']
  },

  // Developer Information
  developer: {
    type: mongoose.Schema.ObjectId,
    ref: 'Developer'
  },
  constructionStatus: {
    type: String,
    enum: [
      'Under Construction',
      'Ready to Move',
      'New Launch',
      'Almost Ready'
    ],
    default: 'Under Construction'
  },
  possessionDate: { type: Date },
  ageOfProperty: {
    type: Number,
    min: 0
  },
  furnished: {
    type: Boolean,
    default: false
  },
  facing: {
    type: String,
    enum: ['North', 'South', 'East', 'West', 'North-East', 'North-West', 'South-East', 'South-West', ''],
    default: ''
  },
  floorRange: {
    min: { type: Number, min: 0 },
    max: { type: Number, min: 0 }
  },
  parkingSpaces: {
    type: Number,
    min: 0,
    default: 0
  },
  verified: {
    type: Boolean,
    default: false
  },
  approvals: [{
    _id: false,
    name: { type: String },
    number: { type: String },
    date: { type: Date }
  }],

  // Location Information
  address: {
    line1: {
      type: String,
      default: '',
      maxlength: [100, 'Address line 1 cannot be more than 100 characters']
    },
    street: {
      type: String,
      required: [true, 'Please add a street address'],
      maxlength: [100, 'Street address cannot be more than 100 characters']
    },
    city: {
      type: String,
      required: [true, 'Please add a city'],
      maxlength: [50, 'City name cannot be more than 50 characters']
    },
    locality: {
      type: String,
      required: [true, 'Please add a locality'],
      maxlength: [50, 'Locality name cannot be more than 50 characters']
    },
    state: {
      type: String,
      required: [true, 'Please add a state'],
      maxlength: [50, 'State name cannot be more than 50 characters']
    },
    zipCode: {
      type: String,
      required: [true, 'Please add a zip code'],
      maxlength: [20, 'Zip code cannot be more than 20 characters'],
      validate: {
        validator: function(v) {
          return /^\d{5,6}(?:[-\s]\d{4})?$/.test(v);
        },
        message: props => `${props.value} is not a valid zip code!`
      }
    },
    country: {
      type: String,
      required: [true, 'Please add a country'],
      default: 'India',
      maxlength: [50, 'Country name cannot be more than 50 characters']
    }
  },
  location: {
    type: {
      type: String,
      enum: ['Point'],
      required: true,
      default: 'Point'
    },
    coordinates: {
      type: [Number],
      required: true,
      default: [0, 0],
      validate: {
        validator: function(v) {
          return v.length === 2 && 
                 v[0] >= -180 && v[0] <= 180 && 
                 v[1] >= -90 && v[1] <= 90;
        },
        message: props => `${props.value} is not a valid coordinate!`
      }
    },
    formattedAddress: String,
    street: String,
    city: String,
    state: String,
    zipCode: String,
    country: String
  },

  // Features and Amenities
  amenities: {
    type: [String],
    default: [],
    enum: [
      'Parking',
      'Swimming Pool',
      'Gym',
      'Security',
      'Garden',
      'Balcony',
      'WiFi',
      'Air Conditioning',
      'Furnished',
      'Pet Friendly',
      'Elevator',
      'Laundry',
      'Storage',
      'Conference Room',
      'Kitchen'
    ]
  },
  highlights: {
    type: [String],
    default: [],
    maxlength: [100, 'Highlight cannot be more than 100 characters']
  },
  nearbyLocalities: {
    hasSchool: { type: Boolean, default: false },
    school: { type: String, default: '' },
    hasHospital: { type: Boolean, default: false },
    hospital: { type: String, default: '' },
    hasMall: { type: Boolean, default: false },
    mall: { type: String, default: '' },
    hasPark: { type: Boolean, default: false },
    park: { type: String, default: '' },
    hasTransport: { type: Boolean, default: false },
    transport: { type: String, default: '' }
  },

  // Project Details
  projectDetails: {
    projectArea: { type: String, default: '' },
    totalUnits: { type: String, default: '' },
    launchDate: { type: Date, default: null },
    reraId: { type: String, default: '' },
    configurations: { type: String, default: '' }
  },

  // Media
  images: [{
    _id: false,
    url: {
      type: String,
      required: true
    },
    publicId: {
      type: String,
      required: true
    },
    width: { type: Number },
    height: { type: Number },
    uploadedAt: {
      type: Date,
      default: Date.now
    }
  }],
  floorPlanImages: [{
    _id: false,
    url: { type: String },
    publicId: { type: String },
    description: { type: String }
  }],
  brochure: {
    url: { type: String },
    publicId: { type: String }
  },
  virtualTour: {
    url: { type: String },
    type: {
      type: String,
      enum: ['video', '3d']
    }
  },

  // Metadata
  featured: {
    type: Boolean,
    default: false
  },
  agent: {
    type: mongoose.Schema.ObjectId,
    ref: 'User',
    required: true
  },
  views: {
    type: Number,
    default: 0
  },
  createdAt: {
    type: Date,
    default: Date.now
  }
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Create property slug from the title
PropertySchema.pre('save', function(next) {
  this.slug = slugify(this.title, { lower: true });
  next();
});

// Geocode & create location field
PropertySchema.pre('save', async function(next) {
  if (!this.isModified('address')) return next();

  try {
    const addressString = [
      this.address.line1,
      this.address.street,
      this.address.city,
      this.address.state,
      this.address.zipCode,
      this.address.country
    ].filter(Boolean).join(', ');

    const loc = await geocoder.geocode(addressString);
    
    if (!loc || loc.length === 0) {
      console.warn('Geocoding failed, using default coordinates');
      return next();
    }

    this.location = {
      type: 'Point',
      coordinates: [loc[0].longitude, loc[0].latitude],
      formattedAddress: loc[0].formattedAddress,
      street: loc[0].streetName || this.address.street,
      city: loc[0].city || this.address.city,
      state: loc[0].stateCode || this.address.state,
      zipCode: loc[0].zipcode || this.address.zipCode,
      country: loc[0].countryCode || this.address.country
    };

    next();
  } catch (err) {
    console.error('Geocoding error:', err);
    next();
  }
});

// Calculate age of property if possession date is provided
PropertySchema.pre('save', function(next) {
  if (this.possessionDate && !this.ageOfProperty) {
    const ageInMilliseconds = Date.now() - new Date(this.possessionDate).getTime();
    this.ageOfProperty = Math.floor(ageInMilliseconds / (1000 * 60 * 60 * 24 * 365));
  }
  next();
});

// Reverse populate with virtuals
PropertySchema.virtual('reviews', {
  ref: 'Review',
  localField: '_id',
  foreignField: 'property',
  justOne: false
});

// Cascade delete reviews when a property is deleted
PropertySchema.pre('deleteOne', { document: true, query: false }, async function(next) {
  await this.model('Review').deleteMany({ property: this._id });
  next();
});

// Index for geospatial queries
PropertySchema.index({ location: '2dsphere' });

// Index for text search (enhanced with weights)
PropertySchema.index({
  title: 'text',
  description: 'text',
  'address.city': 'text',
  'address.state': 'text',
  'address.locality': 'text',
  buildingName: 'text',
  type: 'text'
}, {
  weights: {
    title: 10,
    'address.city': 8,
    'address.locality': 7,
    'address.state': 6,
    buildingName: 5,
    type: 4,
    description: 2
  },
  name: 'text_search_index'
});

// Performance indexes for filtering
PropertySchema.index({ price: 1, area: 1 });
PropertySchema.index({ 'address.city': 1, 'address.locality': 1 });
PropertySchema.index({ constructionStatus: 1 });
PropertySchema.index({ createdAt: -1 });
PropertySchema.index({ views: -1 });
PropertySchema.index({ featured: -1, createdAt: -1 });
PropertySchema.index({ verified: 1 });
PropertySchema.index({ type: 1, status: 1 });
PropertySchema.index({ bedrooms: 1, bathrooms: 1 });

// Compound indexes for common search patterns
PropertySchema.index({ 'address.city': 1, type: 1, status: 1, price: 1 });
PropertySchema.index({ 'address.state': 1, type: 1, status: 1 });
PropertySchema.index({ status: 1, price: 1, area: 1 });
PropertySchema.index({ status: 1, bedrooms: 1, bathrooms: 1 });
PropertySchema.index({ verified: 1, featured: 1, createdAt: -1 });
PropertySchema.index({ 'address.city': 1, price: 1, area: 1 });
PropertySchema.index({ type: 1, price: 1, createdAt: -1 });

// Index for search with filters combination
PropertySchema.index({ 
  'address.city': 1, 
  type: 1, 
  status: 1, 
  price: 1, 
  bedrooms: 1,
  bathrooms: 1
});

// Delete existing model if it exists to prevent caching issues
if (mongoose.models.Property) {
  delete mongoose.models.Property;
}

module.exports = mongoose.model('Property', PropertySchema);

File: mobile/lib/models/property.dart
import 'package:equatable/equatable.dart';

/// Property model
class Property extends Equatable {
  final String id;
  final String title;
  final String description;
  final double price;
  final String currency;
  final String type; // apartment, house, villa, commercial, land
  final String status; // available, sold, rented, under_construction
  final PropertyLocation location;
  final PropertySpecifications specifications;
  final List<String> images;
  final List<String> amenities;
  final PropertyAgent agent;
  final bool featured;
  final int views;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Property({
    required this.id,
    required this.title,
    required this.description,
    required this.price,
    required this.currency,
    required this.type,
    required this.status,
    required this.location,
    required this.specifications,
    required this.images,
    required this.amenities,
    required this.agent,
    required this.featured,
    required this.views,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Property.fromJson(Map<String, dynamic> json) {
    return Property(
      id: json['_id'] ?? json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      currency: json['currency'] ?? 'USD',
      type: json['type'] ?? 'apartment',
      status: json['status'] ?? 'available',
      location: PropertyLocation.fromJson(json['location'] ?? {}),
      specifications: PropertySpecifications.fromJson(json['specifications'] ?? {}),
      images: List<String>.from(json['images'] ?? []),
      amenities: List<String>.from(json['amenities'] ?? []),
      agent: PropertyAgent.fromJson(json['agent'] ?? {}),
      featured: json['featured'] ?? false,
      views: json['views'] ?? 0,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'price': price,
      'currency': currency,
      'type': type,
      'status': status,
      'location': location.toJson(),
      'specifications': specifications.toJson(),
      'images': images,
      'amenities': amenities,
      'agent': agent.toJson(),
      'featured': featured,
      'views': views,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  Property copyWith({
    String? id,
    String? title,
    String? description,
    double? price,
    String? currency,
    String? type,
    String? status,
    PropertyLocation? location,
    PropertySpecifications? specifications,
    List<String>? images,
    List<String>? amenities,
    PropertyAgent? agent,
    bool? featured,
    int? views,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Property(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      price: price ?? this.price,
      currency: currency ?? this.currency,
      type: type ?? this.type,
      status: status ?? this.status,
      location: location ?? this.location,
      specifications: specifications ?? this.specifications,
      images: images ?? this.images,
      amenities: amenities ?? this.amenities,
      agent: agent ?? this.agent,
      featured: featured ?? this.featured,
      views: views ?? this.views,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  List<Object?> get props => [
        id,
        title,
        description,
        price,
        currency,
        type,
        status,
        location,
        specifications,
        images,
        amenities,
        agent,
        featured,
        views,
        createdAt,
        updatedAt,
      ];
}

/// Property location
class PropertyLocation extends Equatable {
  final String address;
  final String city;
  final String state;
  final String country;
  final String zipCode;
  final double latitude;
  final double longitude;
  final String neighborhood;

  const PropertyLocation({
    required this.address,
    required this.city,
    required this.state,
    required this.country,
    required this.zipCode,
    required this.latitude,
    required this.longitude,
    required this.neighborhood,
  });

  factory PropertyLocation.fromJson(Map<String, dynamic> json) {
    return PropertyLocation(
      address: json['address'] ?? '',
      city: json['city'] ?? '',
      state: json['state'] ?? '',
      country: json['country'] ?? '',
      zipCode: json['zipCode'] ?? '',
      latitude: (json['latitude'] ?? 0).toDouble(),
      longitude: (json['longitude'] ?? 0).toDouble(),
      neighborhood: json['neighborhood'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'address': address,
      'city': city,
      'state': state,
      'country': country,
      'zipCode': zipCode,
      'latitude': latitude,
      'longitude': longitude,
      'neighborhood': neighborhood,
    };
  }

  @override
  List<Object?> get props => [
        address,
        city,
        state,
        country,
        zipCode,
        latitude,
        longitude,
        neighborhood,
      ];
}

/// Property specifications
class PropertySpecifications extends Equatable {
  final int bedrooms;
  final int bathrooms;
  final double area; // in square feet/meters
  final String areaUnit; // sqft, sqm
  final int floors;
  final int parkingSpaces;
  final int balconies;
  final String furnishing; // furnished, semi-furnished, unfurnished
  final int age; // in years
  final String facing; // north, south, east, west
  final String floorType; // marble, tiles, wooden, etc.

  const PropertySpecifications({
    required this.bedrooms,
    required this.bathrooms,
    required this.area,
    required this.areaUnit,
    required this.floors,
    required this.parkingSpaces,
    required this.balconies,
    required this.furnishing,
    required this.age,
    required this.facing,
    required this.floorType,
  });

  factory PropertySpecifications.fromJson(Map<String, dynamic> json) {
    return PropertySpecifications(
      bedrooms: json['bedrooms'] ?? 0,
      bathrooms: json['bathrooms'] ?? 0,
      area: (json['area'] ?? 0).toDouble(),
      areaUnit: json['areaUnit'] ?? 'sqft',
      floors: json['floors'] ?? 1,
      parkingSpaces: json['parkingSpaces'] ?? 0,
      balconies: json['balconies'] ?? 0,
      furnishing: json['furnishing'] ?? 'unfurnished',
      age: json['age'] ?? 0,
      facing: json['facing'] ?? 'north',
      floorType: json['floorType'] ?? 'tiles',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'bedrooms': bedrooms,
      'bathrooms': bathrooms,
      'area': area,
      'areaUnit': areaUnit,
      'floors': floors,
      'parkingSpaces': parkingSpaces,
      'balconies': balconies,
      'furnishing': furnishing,
      'age': age,
      'facing': facing,
      'floorType': floorType,
    };
  }

  @override
  List<Object?> get props => [
        bedrooms,
        bathrooms,
        area,
        areaUnit,
        floors,
        parkingSpaces,
        balconies,
        furnishing,
        age,
        facing,
        floorType,
      ];
}

/// Property agent
class PropertyAgent extends Equatable {
  final String id;
  final String name;
  final String email;
  final String phone;
  final String profileImage;
  final String company;
  final double rating;
  final int totalProperties;

  const PropertyAgent({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
    required this.profileImage,
    required this.company,
    required this.rating,
    required this.totalProperties,
  });

  factory PropertyAgent.fromJson(Map<String, dynamic> json) {
    return PropertyAgent(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      phone: json['phone'] ?? '',
      profileImage: json['profileImage'] ?? '',
      company: json['company'] ?? '',
      rating: (json['rating'] ?? 0).toDouble(),
      totalProperties: json['totalProperties'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'email': email,
      'phone': phone,
      'profileImage': profileImage,
      'company': company,
      'rating': rating,
      'totalProperties': totalProperties,
    };
  }

  @override
  List<Object?> get props => [
        id,
        name,
        email,
        phone,
        profileImage,
        company,
        rating,
        totalProperties,
      ];
}

/// Property search filters
class PropertyFilters extends Equatable {
  final String? type;
  final String? status;
  final double? minPrice;
  final double? maxPrice;
  final int? minBedrooms;
  final int? maxBedrooms;
  final int? minBathrooms;
  final int? maxBathrooms;
  final double? minArea;
  final double? maxArea;
  final String? city;
  final String? neighborhood;
  final List<String> amenities;
  final bool featured;

  const PropertyFilters({
    this.type,
    this.status,
    this.minPrice,
    this.maxPrice,
    this.minBedrooms,
    this.maxBedrooms,
    this.minBathrooms,
    this.maxBathrooms,
    this.minArea,
    this.maxArea,
    this.city,
    this.neighborhood,
    this.amenities = const [],
    this.featured = false,
  });

  PropertyFilters copyWith({
    String? type,
    String? status,
    double? minPrice,
    double? maxPrice,
    int? minBedrooms,
    int? maxBedrooms,
    int? minBathrooms,
    int? maxBathrooms,
    double? minArea,
    double? maxArea,
    String? city,
    String? neighborhood,
    List<String>? amenities,
    bool? featured,
  }) {
    return PropertyFilters(
      type: type ?? this.type,
      status: status ?? this.status,
      minPrice: minPrice ?? this.minPrice,
      maxPrice: maxPrice ?? this.maxPrice,
      minBedrooms: minBedrooms ?? this.minBedrooms,
      maxBedrooms: maxBedrooms ?? this.maxBedrooms,
      minBathrooms: minBathrooms ?? this.minBathrooms,
      maxBathrooms: maxBathrooms ?? this.maxBathrooms,
      minArea: minArea ?? this.minArea,
      maxArea: maxArea ?? this.maxArea,
      city: city ?? this.city,
      neighborhood: neighborhood ?? this.neighborhood,
      amenities: amenities ?? this.amenities,
      featured: featured ?? this.featured,
    );
  }

  @override
  List<Object?> get props => [
        type,
        status,
        minPrice,
        maxPrice,
        minBedrooms,
        maxBedrooms,
        minBathrooms,
        maxBathrooms,
        minArea,
        maxArea,
        city,
        neighborhood,
        amenities,
        featured,
      ];
}